dace/frontend/python/preprocessing.py

```diff
@@ -1503,6 +1503,62 @@ def find_disallowed_statements(node: ast.AST):
     return None
 
 
+class MPIResolver(ast.NodeTransformer):
+    """ Resolves mpi4py-related constants, e.g., mpi4py.MPI.COMM_WORLD. """
+    def __init__(self, globals: Dict[str, Any]):
+        from mpi4py import MPI
+        self.globals = globals
+        self.MPI = MPI
+        self.parent = None
+    
+    def visit(self, node):
+        node.parent = self.parent
+        self.parent = node
+        node = super().visit(node)
+        if isinstance(node, ast.AST):
+            self.parent = node.parent
+        return node
+    
+    def visit_Name(self, node: ast.Name) -> Union[ast.Name, ast.Attribute]:
+        self.generic_visit(node)
+        if node.id in self.globals:
+            obj = self.globals[node.id]
+            if isinstance(obj, self.MPI.Comm):
+                lattr = ast.Attribute(ast.Name(id='mpi4py', ctx=ast.Load), attr='MPI')
+                if obj is self.MPI.COMM_NULL:
+                    newnode = ast.copy_location(ast.Attribute(value=lattr, attr='COMM_NULL'), node)
+                    newnode.parent = node.parent
+                    return newnode
+        return node
+    
+    def visit_Attribute(self, node: ast.Attribute) -> ast.Attribute:
+        self.generic_visit(node)
+        if isinstance(node.attr, str) and node.attr == 'Request':
+            try:
+                val = astutils.evalnode(node, self.globals)
+                if val is self.MPI.Request and not isinstance(node.parent, ast.Attribute):
+                    newnode = ast.copy_location(
+                        ast.Attribute(value=ast.Name(id='dace', ctx=ast.Load), attr='MPI_Request'), node)
+                    newnode.parent = node.parent
+                    return newnode
+            except SyntaxError:
+                pass
+        return node
+
+
+class ModuloConverter(ast.NodeTransformer):
+    """ Converts a % b expressions to (a + b) % b for C/C++ compatibility. """
```

Comment:
This is frontend specific (see, e.g., `-2 % 5` Python vs C), so I would definitely **not** handle it in code generation. I don't believe that the added complexity is an issue (sympy is probably not going to simplify the modulo operator anyway, so I doubt it breaks any optimization). We could instead map it to a `pymod` call, but that is probably an even worse solution if the concern is the simplification of symbolic expressions

Reply:
I think this issue comes up now and then, and we forget about it after a while. The problem is that we like to use Python syntax on the SDFG to take advantage of the AST module, but we want this syntax to follow C semantics. Therefore, the frontend has to rewrite certain Python expressions so that they return the same result both in Python and C/C++.

Reply:
Maybe you are thinking that SDFG elements that include Python code should be unparsed to C/C++ equivalent expressions during code generation. However, then the other frontends must change to convert C/Fortran/whatever semantics to Python. Is this a good idea?